node/src/components/chainspec_loader.rs

```diff
@@ -236,6 +247,49 @@ impl ChainspecLoader {
         (chainspec_loader, effects)
     }
 
+    fn handle_initialize<REv>(
+        &mut self,
+        _effect_builder: EffectBuilder<REv>,
+        maybe_highest_block: Option<Box<Block>>,
+    ) -> Effects<Event>
+    where
+        REv: Send,
+    {
+        self.reactor_exit = Some(
+            Self::should_exit_for_upgrade(
+                maybe_highest_block,
+                self.next_upgrade_activation_point(),
+            )
+            .then(|| ReactorExit::ProcessShouldExit(ExitCode::Success))
+            .unwrap_or(ReactorExit::ProcessShouldContinue),
+        );
+        Effects::new()
+    }
+
+    fn should_exit_for_upgrade(
+        maybe_highest_block: Option<Box<Block>>,
+        maybe_next_upgrade_activation_point: Option<ActivationPoint>,
+    ) -> bool {
+        maybe_highest_block.map_or(false, |highest_block| {
+            maybe_next_upgrade_activation_point.map_or(false, |next_upgrade_activation_point| {
+                let highest_block_header = highest_block.header();
+                if highest_block_header.era_id() >= next_upgrade_activation_point.era_id() {
```

Comment:
We could probably exit even if:

```suggestion
                if highest_block_header.next_block_era_id() >= next_upgrade_activation_point.era_id() {
```

Reply:
Yeah, fair point.  Would save us running the chain synchronizer in that case.

Reply:
Fixed in [148a037](https://github.com/casper-network/casper-node/pull/2619/commits/148a037bdc00338d76e60afd30ff1d4834a10b95).